### PR TITLE
expression: Fix wrong way to check for overflow

### LINKED ISSUE #26977: bigint unsigned range error when insert select
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 2. What did you expect to see? (Required)
```sql
mysql> create table t1 (a bigint unsigned);
Query OK, 0 rows affected (0,08 sec)

mysql> insert into t1 values (0xfffffffffffffff0);
Query OK, 1 row affected (0,04 sec)

mysql> insert into t1 select a + 1 from t1;
Query OK, 1 row affected (0,03 sec)
Records: 1  Duplicates: 0  Warnings: 0
```

### 3. What did you see instead (Required)
```sql
tidb> create table t1 (a bigint unsigned);
Query OK, 0 rows affected (0.01 sec)

tidb> insert into t1 values (0xfffffffffffffff0);
Query OK, 1 row affected (0.00 sec)

tidb> insert into t1 select a + 1 from t1;
ERROR 1690 (22003): BIGINT UNSIGNED value is out of range in '(test.t1.a + 1)'
tidb> 
```

### 4. What is your TiDB version? (Required)

Release Version: v5.2.0-alpha-526-g072cf2791-dirty
Edition: Community
Git Commit Hash: 072cf2791e50461f6d9728efc44d184f575a8e81
Git Branch: hexadecimal-literals-23602
UTC Build Time: 2021-08-06 16:34:38
GoVersion: go1.16.6
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by UBarney**
### What problem does this PR solve?

Issue Number: close #26977 

Tests 

- [x] Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by wshwsh12 — APPROVED** [review]

**Comment by UBarney** [comment]
@ichn-hu Is there any problem with this pr ? 😀

**Comment by mjonss** [comment]
Looks good to me, but there have been an update to `expression/integration_test.go` which causes a conflict, @UBarney can you please resolve the conflict and push a new commit? There was also some non-relevant noice in the CI, but that will be rebuilt/retested when you push a new commit.

File: expression/integration_test.go [hunk 1]
```diff
@@ -10208,3 +10208,9 @@ func (s *testIntegrationSuite) TestConstPropNullFunctions(c *C) {
 	tk.MustExec("insert into t2 values (0, 'c', null), (1, null, 0.1), (3, 'b', 0.01), (2, 'q', 0.12), (null, 'a', -0.1), (null, null, null)")
 	tk.MustQuery("select * from t2 where t2.i2=((select count(1) from t1 where t1.i1=t2.i2))").Check(testkit.Rows("1 <nil> 0.1"))
 }
+
+func (s *testIntegrationSuite) TestIssue26977(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	result := tk.MustQuery("select a + 1 as f from (select cast(0xfffffffffffffff0 as unsigned) as a union select cast(1 as unsigned)) t having f != 2;")
+	result.Check(testkit.Rows("18446744073709551601"))
```
Comment (mjonss) [inline expression/integration_test.go:10215]: Could you also add the original test case?

**Review by mjonss — COMMENTED** [review]
Suggesting you to also include the original test case.

File: expression/integration_test.go (hunk 1 above)
Comment (UBarney) [reply expression/integration_test.go:10254]: Do you mean add this to test case ?
```
tidb> create table t1 (a bigint unsigned);
Query OK, 0 rows affected (0.01 sec)

tidb> insert into t1 values (0xfffffffffffffff0);
Query OK, 1 row affected (0.00 sec)

tidb> insert into t1 select a + 1 from t1;
```

I don’t think this is necessary. The test case I added can show that there was a bug before, and this bug has been fixed in this pr

**Review by ichn-hu — APPROVED** [review]
LGTM, soooorry for missing your pin

**Comment by djshow832** [comment]
Please add your release note.

**Comment by UBarney** [comment]
> Please add your release note.

@djshow832 added in PR description

**Pull request merged**